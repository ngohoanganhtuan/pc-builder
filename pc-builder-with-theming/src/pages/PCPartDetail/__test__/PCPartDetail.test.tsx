import { useState } from 'react';
import renderer from 'react-test-renderer';

// Hooks
import { useGetData } from '@hooks';

// Pages
import { PCPartDetail } from '@pages';

// Enums
import { BRAND_NAMES, PC_PART_TYPES } from '@enums';

const mockData = {
  thumbnail: 'https://cdn.techzones.vn/Data/Sites/1/Product/33622/techzones-msi-mag-x570s-tomahawk-max-wifi-4.png',
  name: 'MSI MAG X570S Tomahawk Max WiFi',
  price: 6990000,
  id: 1,
  type: PC_PART_TYPES.MAINBOARD,
  brandName: BRAND_NAMES.MSI,
};

jest.mock('@hooks/useGetData');
jest.mock('react', () => {
  return {
    ...jest.requireActual('react'),
    useState: jest.fn(),
  };
});

const useStateMock = useState as jest.Mock;
const useGetDataMock = useGetData as jest.Mock;

describe('Test [PCPartDetail] page', () => {
  test('Page [PCPartDetail] should render loading spinner when data is fetching', () => {
    useStateMock.mockImplementation(() => [false, jest.fn()]);
    useGetDataMock.mockImplementation(() => ({
      isFetching: true,
      data: mockData,
    }));
    const component = renderer.create(<PCPartDetail />).toJSON();
    expect(component).toMatchSnapshot();
  });

  test('Page [PCPartDetail] should render correctly when data is fetched', () => {
    useStateMock.mockImplementation(() => [false, jest.fn()]);
    useGetDataMock.mockImplementation(() => ({
      isFetching: false,
      data: mockData,
    }));
    const component = renderer.create(<PCPartDetail />).toJSON();
    expect(component).toMatchSnapshot();
  });

  test('Page [PCPartDetail] should render corrrectly when button is clicked', () => {
    useStateMock.mockImplementation(() => [true, jest.fn()]);
    const component = renderer.create(<PCPartDetail />).toJSON();
    expect(component).toMatchSnapshot();
  });
});
