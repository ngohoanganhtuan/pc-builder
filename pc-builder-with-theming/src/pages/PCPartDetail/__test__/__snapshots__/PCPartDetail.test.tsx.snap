// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test [PCPartDetail] page Page [PCPartDetail] should render correctly when data is fetched 1`] = `
<div
  className="pc-part-detail-page container py-5 d-flex flex-column gap-5"
>
  <div
    className="row g-4"
  >
    <div
      className="col-12 col-md-5"
    >
      <img
        className="border rounded pc-part-thumbnail"
        data-testid="pc-part-thumbnail"
        height="100%"
        width="100%"
      />
    </div>
    <div
      className="col-12 col-md-7"
    >
      <h2
        className="name"
      />
      <div
        className="d-flex flex-row justify-content-start gap-3"
      >
        <p
          className="brand text-muted d-flex gap-1 fs-5"
        >
          Brand:
          <span
            className="text-dark fw-bold"
          />
        </p>
        <p
          className="product-id text-muted d-flex gap-1 fs-5"
        >
          Product id:
          <span
            className="text-dark fw-bold"
          />
        </p>
      </div>
      <p
        className="price text-danger fw-bold fs-3"
      >
        NaN ₫
      </p>
      <button
        className="btn btn-lg btn-success"
        disabled={false}
        onClick={[Function]}
      >
        Add this product to my cart
      </button>
    </div>
  </div>
  <table
    className="table table-striped table-bordered border rounded"
  >
    <thead
      className="table-light fw-bold fs-3 pl-3"
    >
      <tr>
        <td
          colSpan={2}
        >
          Specifications
        </td>
      </tr>
    </thead>
    <tbody>
      <tr>
        <th>
          Name
        </th>
        <td />
      </tr>
      <tr>
        <th>
          Hardware type
        </th>
        <td />
      </tr>
      <tr>
        <th>
          Hardware brand
        </th>
        <td />
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`Test [PCPartDetail] page Page [PCPartDetail] should render corrrectly when button is clicked 1`] = `
<div
  className="pc-part-detail-page container py-5 d-flex flex-column gap-5"
>
  <div
    className="row g-4"
  >
    <div
      className="col-12 col-md-5"
    >
      <img
        className="border rounded pc-part-thumbnail"
        data-testid="pc-part-thumbnail"
        height="100%"
        width="100%"
      />
    </div>
    <div
      className="col-12 col-md-7"
    >
      <h2
        className="name"
      />
      <div
        className="d-flex flex-row justify-content-start gap-3"
      >
        <p
          className="brand text-muted d-flex gap-1 fs-5"
        >
          Brand:
          <span
            className="text-dark fw-bold"
          />
        </p>
        <p
          className="product-id text-muted d-flex gap-1 fs-5"
        >
          Product id:
          <span
            className="text-dark fw-bold"
          />
        </p>
      </div>
      <p
        className="price text-danger fw-bold fs-3"
      >
        NaN ₫
      </p>
      <button
        className="btn btn-lg btn-outline-success"
        disabled={true}
        onClick={[Function]}
      >
        Product added to your cart
      </button>
    </div>
  </div>
  <table
    className="table table-striped table-bordered border rounded"
  >
    <thead
      className="table-light fw-bold fs-3 pl-3"
    >
      <tr>
        <td
          colSpan={2}
        >
          Specifications
        </td>
      </tr>
    </thead>
    <tbody>
      <tr>
        <th>
          Name
        </th>
        <td />
      </tr>
      <tr>
        <th>
          Hardware type
        </th>
        <td />
      </tr>
      <tr>
        <th>
          Hardware brand
        </th>
        <td />
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`Test [PCPartDetail] page Page [PCPartDetail] should render loading spinner when data is fetching 1`] = `
<div
  className="spinner-border align-self-center m-auto"
  role="status"
/>
`;
